docs/guide/interactions/views.rst

```diff
@@ -0,0 +1,483 @@
+            component: discord.ui.Select['RoleSelector'],
+        ) -> None:
+            raise NotImplementedError
+
+We also need to add a body to our ``selector`` callback function, which will assign the selected role to the user.
+This is fairly simple to do, when the callback is invoked we can access the options the user selected via the
+:attr:`~discord.ui.Select.options` attribute. This holds a list of values, but since we limited the number of values to `1`,
+we can just access the first element directly. We can then use sets to determine what roles the user already has, 
+and then override the member's roles with :meth:`Member.edit <discord.Member.edit>`.
+
+Our function body might look like this:
+
+.. code-block:: python
+    :emphasize-lines: 21-25
+
+    class RoleSelector(discord.ui.View):
+        def __init__(self, message_id: int, roles: List[discord.Role]) -> None:
+            super().__init__(timeout=None)
+
+            self.selector.custom_id = f'role_selector:{message_id}'
+
+            self.roles: List[discord.Role] = roles
+            for role in roles:
+                self.selector.add_option(label=role.name, value=str(role.id))
+
+        @discord.ui.select(
+            placeholder='Select a role',
+            min_values=1,
+            max_values=1,
+        )
+        async def selector(
+            self,
+            interaction: discord.Interaction,
+            component: discord.ui.Select['RoleSelector'],
+        ) -> None:
+            role_id = int(component.values[0])
+            role = discord.Object(id=role_id, type=discord.Role)
+            roles = set(interaction.user.roles) - set(self.roles) | {role}
+            await interaction.user.edit(roles=roles)
+            await interaction.response.defer()
+
+
+Making a View Persist
+~~~~~~~~~~~~~~~~~~~~~~
+
+Now our view class is complete, we can manually attach an instance to a message to retrieve the ID.
+When we're doing this, the ``custom_id`` can be set to anything since the view handling will be done 
+when we mark the view as persistent.
+
+.. code-block:: python
+
+    ROLE_IDS = [490320652230852629, 714516281293799438, 859169678966784031, 673195396834656257]
+    roles = [guild.get_role(id) for id in ROLE_IDS]
+    await channel.send('Choose your house', view=RoleSelector(0, roles)) # we don't know the message ID yet.
+
+Once we have a message ID, we just need to create an instance of our view and attach it to the :class:`~discord.Client`
+with the :meth:`~discord.Client.add_view` method. This is done in the ``setup_hook`` method of our :class:`~discord.Client` 
+subclass to ensure the view is attached on startup.
+
+.. code-block:: python
+
+    GUILD_ID = 336642139381301249
+    SELECTOR_MESSAGE_ID = 881049369850306610
+    ROLE_IDS = [490320652230852629, 714516281293799438, 859169678966784031, 673195396834656257]
+
+    class MyClient(discord.Client):
+
+        async def setup_role_selector(self) -> None:
```

Comment:
It will not block **because** of the `asyncio.create_task()` in the `setup_hook()`
The warning in the docs is for the case of doing the `wait_until_ready()` in the `setup_hook()` directly.
`asyncio.create_task()` doesn't block the function from continuing like  `wait_for` and `wait_until_ready()` do, it schedules the task on the loop and then continues which prevents the deadlock that would be caused by a `wait_until_ready()`
Test the code yourself and you will see it doesn't deadlock the bot
or here a more minimal example showing that it won't deadlock
```py
#!/usr/bin/env python3

from discord.ext import commands
import discord
import config
import asyncio

class Bot(commands.Bot):
    def __init__(self, intents: discord.Intents, **kwargs):
        super().__init__(command_prefix=commands.when_mentioned_or('$'), intents=intents, **kwargs)

    async def ready_waiter(self):
        print("waiting for ready event....")
        await self.wait_until_ready()
        print("ready!!!!")

    async def setup_hook(self):
        asyncio.create_task(self.ready_waiter())

    async def on_ready(self):
        print(f'Logged on as {self.user} (ID: {self.user.id})')


intents = discord.Intents.default()
intents.message_content = True
bot = Bot(intents=intents)

# write general commands here

bot.run(config.token)

```

Should show the following output:
```
(.venv) ➜  SetupBot python .\bot.py
2024-02-06 10:48:33 INFO     discord.client logging in using static token
waiting for ready event....
2024-02-06 10:48:34 INFO     discord.gateway Shard ID None has connected to Gateway (Session ID: e40b991f53a805e1a5f4a9148a1509a7).
ready!!!!    
```

The way the guide is doing is fine and actually helpful since it shows how to make a setup part that needs to wait for the cache to be ready, to add the view, which is a bit more complex than the linked example in your comment, which just uses static values for the sake of simplicity and showing what is needed for a persistent view (which means it doesn't need a ready cache) vs. what this example here is doing of showing what to do with more dynamic parts in a persistent view and objects needing to be in cache to actually add the view.

Reply:
@tailoric That is perfectly understandable, but *why* wait until the internal caches have been set in order to then add the view? Unless the role selector view depends on some internal cache to be set, I don't really see an purpose of using `wait_until_ready` in this context

Reply:
It does need the cache.
The view relies on a list of roles to get the name and id of the role to add to the select options.
See Lines [277-279 ](https://github.com/Rapptz/discord.py/blob/8a3e3e56e3e7eec10e680cd7f07628384b6217ed/docs/guide/interactions/views.rst?plain=1#L277-L279)this very file. That's why the `wait_until_ready` is needed in this example since this is supposed to be a permanent dropdown with self assignable roles.

Reply:
> wait_until_ready is needed in this example since this is supposed to be a permanent dropdown with self assignable roles.

@tailoric `wait_until_ready` is required because of the guild cache not being set within `setup_hook`. I have validated and tested this code on dpy 2.3.2 using Python 3.12.1 and yes it does require the cache to be set. The reason is not what you described, but rather because the guild is `None` as no data is populated within the cache at that time. Thus it errors out.

Reply:
I don't think what I said was wrong here.
I said the cache needs to be ready to get the roles, which yes in turn you also need to do `get_guild` which also has to be cached so you can use `get_role` on the guild to get the role instances for the list that will be provided to the `__init__` of the view.
But I am also not interested into getting into unnecessary pedant-isms anyway.

The main takeaway and the reason for my reply to your initial review comment was to show and tell you that the `wait_until_ready` was necessary and your suggestion of moving the code directly into the `setup_hook` was not possible with how the view is currently constructed.
I also see the value in showing how to defer the creation of a persistent view, that needs some bot cache objects to be ready, before being re-added to the view store, after a restart, which this example shows off. 

This conversation however also showed that it may be beneficial to elaborate why this example is putting the initialization of the `RoleSelector` view into an `asyncio.create_task` to avoid the deadlock mentioned in the documentation.

Reply:
I agree with in this case. I'll leave it as is